src/index/mod.rs

```diff
@@ -0,0 +1,160 @@
+//! ```no_run
+//! use pravega_client_macros::Label;
+//!
+//! #[derive(Label, Debug, PartialOrd, PartialEq)]
+//! struct MyLabel {
+//!     time: u64,
+//!     id: u64,
+//! }
+//!
+//! ```
+//! The `Label` procedural marco will convert the `MyLabel` struct to a list of key-value pairs.
+//! Notice that we only accept u64 as the `Entry` value type.
+//!
+//! To ensure the searching efficiency, we impose some constraints to the `Label`:
+//! * The `Entry` value in the `Label` must be monotonically increasing.
+//! * Deletion of the `Entry` in the `Label` is not permitted.
+//! * Adding a new `Entry` is possible, but the `Entry` has to be appended at the tail.
+//! * The Index Writer is generic over the `Label` struct, meaning if a new `Entry` is appended, it needs
+//! to create a new Index Writer for the new `Label`.
+//!
+//! [`Record`]: crate::index::Record
+
+pub mod writer;
+#[doc(inline)]
+pub use writer::IndexWriter;
+
+pub mod reader;
+#[doc(inline)]
+pub use reader::IndexReader;
+
+use bincode2::Config;
+use bincode2::Error as BincodeError;
+use bincode2::LengthOption;
+use lazy_static::*;
+use pravega_wire_protocol::commands::{Command, EventCommand};
+use serde::{Deserialize, Serialize};
+use tiny_keccak::{Hasher, Shake};
+
+const RECORD_SIZE: u64 = 4 * 1024;
+const DATA_SIZE: usize = 1024;
+
+lazy_static! {
+    static ref CONFIG: Config = {
+        let mut config = bincode2::config();
+        config.big_endian();
+        config.limit(RECORD_SIZE);
+        config.array_length(LengthOption::U32);
+        config.string_length(LengthOption::U16);
+        config
+    };
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Record {
+    type_code: i32,
+    entries_len: u32,
+    data_len: u32,
+    entries: Vec<(u128, u64)>,
+    data: Vec<u8>,
+}
+
+impl Record {
+    const TYPE_CODE: i32 = 0;
+
+    pub(crate) fn new(entries: Vec<(u128, u64)>, data: Vec<u8>) -> Record {
+        Record {
+            type_code: Record::TYPE_CODE,
+            // u128 is 16 bytes and u64 is 8 bytes
+            entries_len: entries.len() as u32 * 24,
+            data_len: data.len() as u32,
+            entries,
+            data,
+        }
+    }
+
+    fn write_fields(&self) -> Result<Vec<u8>, BincodeError> {
+        let mut res = vec![];
+        res.extend_from_slice(&EventCommand::TYPE_CODE.to_be_bytes());
+        res.extend_from_slice(&((RECORD_SIZE - 8) as i32).to_be_bytes());
+        let encoded = CONFIG.serialize(&self)?;
+        let length = encoded.len();
+        res.extend(encoded);
+        let padding = vec![0u8; RECORD_SIZE as usize - length - 8];
+        res.extend(padding);
+        Ok(res)
+    }
+
+    fn read_from(input: &[u8]) -> Result<Self, BincodeError> {
+        let decoded: Record = CONFIG.deserialize(&input[8..])?;
+        Ok(decoded)
+    }
+}
+
+pub(crate) fn hash_key_to_u128(key: &'static str) -> u128 {
+    let mut shake = Shake::v128();
```

Comment:
Can you share the doc link for this library here?

Reply:
Sure, https://docs.rs/tiny-keccak/2.0.2/tiny_keccak/struct.Shake.html as Tom suggests https://github.com/pravega/pravega-client-rust/issues/234#issuecomment-854095681